integration/simulation/simulation_test.go

```diff
@@ -96,67 +139,58 @@ func validateL1(t *testing.T, b *common.Block, s *Stats, db enclave2.DB) {
-		t.Errorf("Deposit amounts don't match. Found %d , expected %d", totalDeposited, s.totalDepositedAmount)
+	if totalDeposited != stats.totalDepositedAmount {
+		t.Errorf("Deposit amounts don't match. Found %d , expected %d", totalDeposited, stats.totalDepositedAmount)
 	}
 
-	efficiency := float64(s.totalL1Blocks-s.l1Height) / float64(s.totalL1Blocks)
+	efficiency := float64(stats.totalL1Blocks-stats.l1Height) / float64(stats.totalL1Blocks)
 	if efficiency > L1EfficiencyThreashold {
 		t.Errorf("Efficiency in L1 is %f. Expected:%f", efficiency, L1EfficiencyThreashold)
 	}
 
 	// todo
-	for nodeID, reorgs := range s.noL1Reorgs {
-		eff := float64(reorgs) / float64(s.l1Height)
+	for nodeID, reorgs := range stats.noL1Reorgs {
+		eff := float64(reorgs) / float64(stats.l1Height)
 		if eff > L1EfficiencyThreashold {
 			t.Errorf("Efficiency for node %d in L1 is %f. Expected:%f", nodeID, eff, L1EfficiencyThreashold)
 		}
 	}
 }
 
-func validateL2(t *testing.T, r *enclave2.Rollup, s *Stats, db enclave2.DB) uint64 {
-	s.l2Height = db.Height(r)
-	transfers := make([]uuid.UUID, 0)
-	withdrawalTxs := make([]enclave2.L2Tx, 0)
-	withdrawalRequests := make([]obscuroCommon.Withdrawal, 0)
-	for {
-		if db.Height(r) == common.L2GenesisHeight {
-			break
-		}
-		for _, tx := range r.Transactions {
-			switch tx.TxType {
-			case enclave2.TransferTx:
-				transfers = append(transfers, tx.ID)
-			case enclave2.WithdrawalTx:
-				withdrawalTxs = append(withdrawalTxs, tx)
-			default:
-				panic("Invalid tx type")
-			}
-		}
-		withdrawalRequests = append(withdrawalRequests, r.Header.Withdrawals...)
-		r = db.Parent(r)
-	}
+func validateL2Stats(t *testing.T, stats *Stats, l2Height int, txManager *TransactionManager) uint64 {
+	// get non-failed transactions
+	transactions := txManager.GetL2Transactions()
+	withdrawalRequests := txManager.GetL2WithdrawalRequests()
+
 	// todo - check that proofs are on the canonical chain
 
-	if len(common.FindDups(transfers)) > 0 {
-		dups := common.FindDups(transfers)
-		t.Errorf("Found L2 txs duplicates: %v", dups)
+	// Sum all withdraw transactions AND
+	txSumWithdrawalsRequested := sumWithdrawals(withdrawalRequests)
+
+	// are there duplicated transaction Ids ?
+	txIds := make([]uuid.UUID, len(transactions))
+	for i, transaction := range transactions {
+		txIds[i] = transaction.ID
 	}
-	if len(transfers) != s.nrTransferTransactions {
-		t.Errorf("Nr of transfers don't match. Found %d , expected %d", len(transfers), s.nrTransferTransactions)
+	if dups := common.FindDups(txIds); len(dups) > 0 {
+		t.Errorf("Found L2 txs duplicates: %v", dups)
 	}
-	if sumWithdrawalTxs(withdrawalTxs) != s.totalWithdrawnAmount {
-		t.Errorf("Withdrawal tx amounts don't match. Found %d , expected %d", sumWithdrawalTxs(withdrawalTxs), s.totalWithdrawnAmount)
+
+	// are the number of txs sent by the generator the same as the captured by the enclave stats ?
+	if len(transactions) != stats.nrTransferTransactions {
```

Comment:
stats.nrTransferTransactions is no longer doing anything since you're capturing the transactions. This check can be removed